From b73c4f6e5ad772aca5a8990f0ccc45d387ee7ac6 Mon Sep 17 00:00:00 2001
From: Hou Zhiqiang <Zhiqiang.Hou@nxp.com>
Date: Mon, 9 Sep 2019 16:04:02 +0800
Subject: [PATCH 06/10] lx2160a: Add workaround for erratum A-050234

The elastic buffer integrated in the receiver path of PCI Express
controller is used to compensate the worst-case frequency difference
between link partners on each side of the PCI Express link by adding
or removing SKP Symbols (part of SKIP Ordered Sets) toward the
received serial data stream.

Due to this erratum, the default configuration of the elastic buffer
thresholds in the PCI Express controller is not optimal and can cause
various link errors.

Signed-off-by: Hou Zhiqiang <Zhiqiang.Hou@nxp.com>
---
 .../rcw_1900_600_2600_13_5_2.rcw              |  1 +
 .../rcw_2000_700_2900_13_5_2.rcw              |  1 +
 .../rcw_2000_800_2900_13_5_2.rcw              |  3 ++
 .../rcw_2000_700_2600_13_3_2.rcw              |  1 +
 .../rcw_2000_800_2600_13_3_2.rcw              |  1 +
 .../rcw_2000_700_2600_13_3_3.rcw              |  1 +
 .../rcw_2000_800_2600_13_3_3.rcw              |  1 +
 .../rcw_2000_700_2900_14_5_2.rcw              |  1 +
 .../rcw_2000_800_2900_14_5_2.rcw              |  1 +
 .../rcw_2000_700_2600_8_3_2.rcw               |  1 +
 .../rcw_2000_700_2600_8_3_3.rcw               |  1 +
 .../rcw_2000_700_2900_3_3_2.rcw               |  1 +
 .../rcw_1900_600_2600_7_5_2.rcw               |  1 +
 .../rcw_2000_700_2400_7_5_2.rcw               |  1 +
 .../rcw_2000_700_2900_7_5_2.rcw               |  1 +
 .../rcw_1900_600_2600_7_3_3.rcw               |  1 +
 .../rcw_1900_600_2600_7_3_3_dspi.rcw          |  1 +
 .../rcw_2000_700_2900_18_2_2.rcw              |  1 +
 .../rcw_2000_700_2600_19_5_2_sd.rcw           |  1 +
 .../rcw_1900_600_2600_15_5_2.rcw              |  1 +
 .../rcw_2000_700_2900_15_5_2.rcw              |  1 +
 .../rcw_2000_700_2400_1_2_2.rcw               |  1 +
 .../rcw_2000_700_2600_1_2_2.rcw               |  1 +
 .../rcw_2000_700_2900_1_2_2.rcw               |  1 +
 lx2160aqds/README                             |  7 ++++
 .../rcw_2000_700_2400_20_11_3.rcw             |  1 +
 .../rcw_2000_700_2900_18_5_2.rcw              |  1 +
 lx2160ardb/README                             |  7 ++++
 .../rcw_1900_600_1600_19_5_2.rcw              |  1 +
 .../rcw_1900_600_2400_19_5_2.rcw              |  1 +
 .../rcw_1900_600_2600_19_5_2.rcw              |  1 +
 .../rcw_1900_600_2600_19_5_2_sd.rcw           |  1 +
 .../rcw_2000_700_2400_19_5_2.rcw              |  1 +
 .../rcw_2000_700_2900_19_5_2.rcw              |  1 +
 .../rcw_2000_700_2900_19_5_2_sd.rcw           |  1 +
 lx2160asi/a050234.rcw                         | 33 +++++++++++++++++++
 36 files changed, 82 insertions(+)
 create mode 100644 lx2160asi/a050234.rcw

diff --git a/lx2160aqds/CCCC_CCCC_PPPP_HHHH_PPPP_PPPP_RR_13_5_2/rcw_1900_600_2600_13_5_2.rcw b/lx2160aqds/CCCC_CCCC_PPPP_HHHH_PPPP_PPPP_RR_13_5_2/rcw_1900_600_2600_13_5_2.rcw
index da4c12f..50d1158 100644
--- a/lx2160aqds/CCCC_CCCC_PPPP_HHHH_PPPP_PPPP_RR_13_5_2/rcw_1900_600_2600_13_5_2.rcw
+++ b/lx2160aqds/CCCC_CCCC_PPPP_HHHH_PPPP_PPPP_RR_13_5_2/rcw_1900_600_2600_13_5_2.rcw
@@ -55,6 +55,7 @@ SRDS_DIV_PEX_S3=1
 
 /* Errata for PCIe controller */
 #include <../lx2160asi/a011270.rcw>
+#include <../lx2160asi/a050234.rcw>
 
 /* common PBI commands */
 #include <../lx2160asi/common.rcw>
diff --git a/lx2160aqds/CCCC_CCCC_PPPP_HHHH_PPPP_PPPP_RR_13_5_2/rcw_2000_700_2900_13_5_2.rcw b/lx2160aqds/CCCC_CCCC_PPPP_HHHH_PPPP_PPPP_RR_13_5_2/rcw_2000_700_2900_13_5_2.rcw
index e41e869..fb13004 100644
--- a/lx2160aqds/CCCC_CCCC_PPPP_HHHH_PPPP_PPPP_RR_13_5_2/rcw_2000_700_2900_13_5_2.rcw
+++ b/lx2160aqds/CCCC_CCCC_PPPP_HHHH_PPPP_PPPP_RR_13_5_2/rcw_2000_700_2900_13_5_2.rcw
@@ -55,6 +55,7 @@ SRDS_DIV_PEX_S3=1
 
 /* Errata for PCIe controller */
 #include <../lx2160asi/a011270.rcw>
+#include <../lx2160asi/a050234.rcw>
 
 /* common PBI commands */
 #include <../lx2160asi/common.rcw>
diff --git a/lx2160aqds/CCCC_CCCC_PPPP_HHHH_PPPP_PPPP_RR_13_5_2/rcw_2000_800_2900_13_5_2.rcw b/lx2160aqds/CCCC_CCCC_PPPP_HHHH_PPPP_PPPP_RR_13_5_2/rcw_2000_800_2900_13_5_2.rcw
index 4a1e163..df851cf 100644
--- a/lx2160aqds/CCCC_CCCC_PPPP_HHHH_PPPP_PPPP_RR_13_5_2/rcw_2000_800_2900_13_5_2.rcw
+++ b/lx2160aqds/CCCC_CCCC_PPPP_HHHH_PPPP_PPPP_RR_13_5_2/rcw_2000_800_2900_13_5_2.rcw
@@ -49,5 +49,8 @@ SRDS_DIV_PEX_S3=1
 /* Modify FlexSPI Clock Divisor value */
 #include <../lx2160asi/flexspi_divisor_28.rcw>
 
+/* Errata for PCIe controller */
+#include <../lx2160asi/a050234.rcw>
+
 /* common PBI commands */
 #include <../lx2160asi/common.rcw>
diff --git a/lx2160aqds/CCCC_CCCC_PPPP_PPPP_PPPP_PPPP_RR_13_3_2/rcw_2000_700_2600_13_3_2.rcw b/lx2160aqds/CCCC_CCCC_PPPP_PPPP_PPPP_PPPP_RR_13_3_2/rcw_2000_700_2600_13_3_2.rcw
index c1268c9..ad70fe8 100644
--- a/lx2160aqds/CCCC_CCCC_PPPP_PPPP_PPPP_PPPP_RR_13_3_2/rcw_2000_700_2600_13_3_2.rcw
+++ b/lx2160aqds/CCCC_CCCC_PPPP_PPPP_PPPP_PPPP_RR_13_3_2/rcw_2000_700_2600_13_3_2.rcw
@@ -53,6 +53,7 @@ SRDS_DIV_PEX_S3=1
 
 /* Errata for PCIe controller */
 #include <../lx2160asi/a011270.rcw>
+#include <../lx2160asi/a050234.rcw>
 
 /* common PBI commands */
 #include <../lx2160asi/common.rcw>
diff --git a/lx2160aqds/CCCC_CCCC_PPPP_PPPP_PPPP_PPPP_RR_13_3_2/rcw_2000_800_2600_13_3_2.rcw b/lx2160aqds/CCCC_CCCC_PPPP_PPPP_PPPP_PPPP_RR_13_3_2/rcw_2000_800_2600_13_3_2.rcw
index 3e4ff67..f275764 100644
--- a/lx2160aqds/CCCC_CCCC_PPPP_PPPP_PPPP_PPPP_RR_13_3_2/rcw_2000_800_2600_13_3_2.rcw
+++ b/lx2160aqds/CCCC_CCCC_PPPP_PPPP_PPPP_PPPP_RR_13_3_2/rcw_2000_800_2600_13_3_2.rcw
@@ -53,6 +53,7 @@ SRDS_DIV_PEX_S3=1
 
 /* Errata for PCIe controller */
 #include <../lx2160asi/a011270.rcw>
+#include <../lx2160asi/a050234.rcw>
 
 /* common PBI commands */
 #include <../lx2160asi/common.rcw>
diff --git a/lx2160aqds/CCCC_CCCC_PPPP_PPPP_PPPP_PPPP_RR_13_3_3/rcw_2000_700_2600_13_3_3.rcw b/lx2160aqds/CCCC_CCCC_PPPP_PPPP_PPPP_PPPP_RR_13_3_3/rcw_2000_700_2600_13_3_3.rcw
index 22395ed..207188f 100644
--- a/lx2160aqds/CCCC_CCCC_PPPP_PPPP_PPPP_PPPP_RR_13_3_3/rcw_2000_700_2600_13_3_3.rcw
+++ b/lx2160aqds/CCCC_CCCC_PPPP_PPPP_PPPP_PPPP_RR_13_3_3/rcw_2000_700_2600_13_3_3.rcw
@@ -53,6 +53,7 @@ SRDS_DIV_PEX_S3=1
 
 /* Errata for PCIe controller */
 #include <../lx2160asi/a011270.rcw>
+#include <../lx2160asi/a050234.rcw>
 
 /* common PBI commands */
 #include <../lx2160asi/common.rcw>
diff --git a/lx2160aqds/CCCC_CCCC_PPPP_PPPP_PPPP_PPPP_RR_13_3_3/rcw_2000_800_2600_13_3_3.rcw b/lx2160aqds/CCCC_CCCC_PPPP_PPPP_PPPP_PPPP_RR_13_3_3/rcw_2000_800_2600_13_3_3.rcw
index d85757a..0e8a272 100644
--- a/lx2160aqds/CCCC_CCCC_PPPP_PPPP_PPPP_PPPP_RR_13_3_3/rcw_2000_800_2600_13_3_3.rcw
+++ b/lx2160aqds/CCCC_CCCC_PPPP_PPPP_PPPP_PPPP_RR_13_3_3/rcw_2000_800_2600_13_3_3.rcw
@@ -53,6 +53,7 @@ SRDS_DIV_PEX_S3=1
 
 /* Errata for PCIe controller */
 #include <../lx2160asi/a011270.rcw>
+#include <../lx2160asi/a050234.rcw>
 
 /* common PBI commands */
 #include <../lx2160asi/common.rcw>
diff --git a/lx2160aqds/CCCC_PPPP_PPPP_HHHH_PPPP_PPPP_RR_14_5_2/rcw_2000_700_2900_14_5_2.rcw b/lx2160aqds/CCCC_PPPP_PPPP_HHHH_PPPP_PPPP_RR_14_5_2/rcw_2000_700_2900_14_5_2.rcw
index e36bd05..6f4b70e 100644
--- a/lx2160aqds/CCCC_PPPP_PPPP_HHHH_PPPP_PPPP_RR_14_5_2/rcw_2000_700_2900_14_5_2.rcw
+++ b/lx2160aqds/CCCC_PPPP_PPPP_HHHH_PPPP_PPPP_RR_14_5_2/rcw_2000_700_2900_14_5_2.rcw
@@ -50,6 +50,7 @@ SRDS_DIV_PEX_S3=1
 
 /* Errata for PCIe controller */
 #include <../lx2160asi/a011270.rcw>
+#include <../lx2160asi/a050234.rcw>
 
 /* common PBI commands */
 #include <../lx2160asi/common.rcw>
diff --git a/lx2160aqds/CCCC_PPPP_PPPP_HHHH_PPPP_PPPP_RR_14_5_2/rcw_2000_800_2900_14_5_2.rcw b/lx2160aqds/CCCC_PPPP_PPPP_HHHH_PPPP_PPPP_RR_14_5_2/rcw_2000_800_2900_14_5_2.rcw
index 385ca8a..54970a8 100644
--- a/lx2160aqds/CCCC_PPPP_PPPP_HHHH_PPPP_PPPP_RR_14_5_2/rcw_2000_800_2900_14_5_2.rcw
+++ b/lx2160aqds/CCCC_PPPP_PPPP_HHHH_PPPP_PPPP_RR_14_5_2/rcw_2000_800_2900_14_5_2.rcw
@@ -50,6 +50,7 @@ SRDS_DIV_PEX_S3=1
 
 /* Errata for PCIe controller */
 #include <../lx2160asi/a011270.rcw>
+#include <../lx2160asi/a050234.rcw>
 
 /* common PBI commands */
 #include <../lx2160asi/common.rcw>
diff --git a/lx2160aqds/FFFF_FFFF_PPPP_PPPP_PPPP_PPPP_RR_8_3_2/rcw_2000_700_2600_8_3_2.rcw b/lx2160aqds/FFFF_FFFF_PPPP_PPPP_PPPP_PPPP_RR_8_3_2/rcw_2000_700_2600_8_3_2.rcw
index 7c43f5c..645d888 100644
--- a/lx2160aqds/FFFF_FFFF_PPPP_PPPP_PPPP_PPPP_RR_8_3_2/rcw_2000_700_2600_8_3_2.rcw
+++ b/lx2160aqds/FFFF_FFFF_PPPP_PPPP_PPPP_PPPP_RR_8_3_2/rcw_2000_700_2600_8_3_2.rcw
@@ -54,6 +54,7 @@ SRDS_DIV_PEX_S3=1
 
 /* Errata for PCIe controller */
 #include <../lx2160asi/a011270.rcw>
+#include <../lx2160asi/a050234.rcw>
 
 /* common PBI commands */
 #include <../lx2160asi/common.rcw>
diff --git a/lx2160aqds/FFFF_FFFF_PPPP_PPPP_PPPP_PPPP_RR_8_3_3/rcw_2000_700_2600_8_3_3.rcw b/lx2160aqds/FFFF_FFFF_PPPP_PPPP_PPPP_PPPP_RR_8_3_3/rcw_2000_700_2600_8_3_3.rcw
index e46900d..a3ba53d 100644
--- a/lx2160aqds/FFFF_FFFF_PPPP_PPPP_PPPP_PPPP_RR_8_3_3/rcw_2000_700_2600_8_3_3.rcw
+++ b/lx2160aqds/FFFF_FFFF_PPPP_PPPP_PPPP_PPPP_RR_8_3_3/rcw_2000_700_2600_8_3_3.rcw
@@ -53,6 +53,7 @@ SRDS_DIV_PEX_S3=1
 
 /* Errata for PCIe controller */
 #include <../lx2160asi/a011270.rcw>
+#include <../lx2160asi/a050234.rcw>
 
 /* common PBI commands */
 #include <../lx2160asi/common.rcw>
diff --git a/lx2160aqds/FFFF_PPPP_PPPP_PPPP_PPPP_PPPP_RR_3_3_2/rcw_2000_700_2900_3_3_2.rcw b/lx2160aqds/FFFF_PPPP_PPPP_PPPP_PPPP_PPPP_RR_3_3_2/rcw_2000_700_2900_3_3_2.rcw
index c629bc5..b8a6315 100644
--- a/lx2160aqds/FFFF_PPPP_PPPP_PPPP_PPPP_PPPP_RR_3_3_2/rcw_2000_700_2900_3_3_2.rcw
+++ b/lx2160aqds/FFFF_PPPP_PPPP_PPPP_PPPP_PPPP_RR_3_3_2/rcw_2000_700_2900_3_3_2.rcw
@@ -53,6 +53,7 @@ SRDS_PLL_REF_CLK_SEL_S1=2
 
 /* Errata for PCIe controller */
 #include <../lx2160asi/a011270.rcw>
+#include <../lx2160asi/a050234.rcw>
 
 /* common PBI commands */
 #include <../lx2160asi/common.rcw>
diff --git a/lx2160aqds/FFFF_SSSS_PPPP_HHHH_PPPP_PPPP_RR_7_5_2/rcw_1900_600_2600_7_5_2.rcw b/lx2160aqds/FFFF_SSSS_PPPP_HHHH_PPPP_PPPP_RR_7_5_2/rcw_1900_600_2600_7_5_2.rcw
index 8300c6a..67f31bf 100644
--- a/lx2160aqds/FFFF_SSSS_PPPP_HHHH_PPPP_PPPP_RR_7_5_2/rcw_1900_600_2600_7_5_2.rcw
+++ b/lx2160aqds/FFFF_SSSS_PPPP_HHHH_PPPP_PPPP_RR_7_5_2/rcw_1900_600_2600_7_5_2.rcw
@@ -55,6 +55,7 @@ SRDS_DIV_PEX_S3=1
 
 /* Errata for PCIe controller */
 #include <../lx2160asi/a011270.rcw>
+#include <../lx2160asi/a050234.rcw>
 
 /* common PBI commands */
 #include <../lx2160asi/common.rcw>
diff --git a/lx2160aqds/FFFF_SSSS_PPPP_HHHH_PPPP_PPPP_RR_7_5_2/rcw_2000_700_2400_7_5_2.rcw b/lx2160aqds/FFFF_SSSS_PPPP_HHHH_PPPP_PPPP_RR_7_5_2/rcw_2000_700_2400_7_5_2.rcw
index 351dc76..3ca14de 100644
--- a/lx2160aqds/FFFF_SSSS_PPPP_HHHH_PPPP_PPPP_RR_7_5_2/rcw_2000_700_2400_7_5_2.rcw
+++ b/lx2160aqds/FFFF_SSSS_PPPP_HHHH_PPPP_PPPP_RR_7_5_2/rcw_2000_700_2400_7_5_2.rcw
@@ -55,6 +55,7 @@ SRDS_DIV_PEX_S3=1
 
 /* Errata for PCIe controller */
 #include <../lx2160asi/a011270.rcw>
+#include <../lx2160asi/a050234.rcw>
 
 /* common PBI commands */
 #include <../lx2160asi/common.rcw>
diff --git a/lx2160aqds/FFFF_SSSS_PPPP_HHHH_PPPP_PPPP_RR_7_5_2/rcw_2000_700_2900_7_5_2.rcw b/lx2160aqds/FFFF_SSSS_PPPP_HHHH_PPPP_PPPP_RR_7_5_2/rcw_2000_700_2900_7_5_2.rcw
index 74fdc8c..c674df1 100644
--- a/lx2160aqds/FFFF_SSSS_PPPP_HHHH_PPPP_PPPP_RR_7_5_2/rcw_2000_700_2900_7_5_2.rcw
+++ b/lx2160aqds/FFFF_SSSS_PPPP_HHHH_PPPP_PPPP_RR_7_5_2/rcw_2000_700_2900_7_5_2.rcw
@@ -55,6 +55,7 @@ SRDS_DIV_PEX_S3=1
 
 /* Errata for PCIe controller */
 #include <../lx2160asi/a011270.rcw>
+#include <../lx2160asi/a050234.rcw>
 
 /* common PBI commands */
 #include <../lx2160asi/common.rcw>
diff --git a/lx2160aqds/FFFF_SSSS_PPPP_PPPP_PPPP_PPPP_RR_7_3_3/rcw_1900_600_2600_7_3_3.rcw b/lx2160aqds/FFFF_SSSS_PPPP_PPPP_PPPP_PPPP_RR_7_3_3/rcw_1900_600_2600_7_3_3.rcw
index e3922f5..b8e845e 100644
--- a/lx2160aqds/FFFF_SSSS_PPPP_PPPP_PPPP_PPPP_RR_7_3_3/rcw_1900_600_2600_7_3_3.rcw
+++ b/lx2160aqds/FFFF_SSSS_PPPP_PPPP_PPPP_PPPP_RR_7_3_3/rcw_1900_600_2600_7_3_3.rcw
@@ -59,6 +59,7 @@ SRDS_DIV_PEX_S3=1
 
 /* Errata for PCIe controller */
 #include <../lx2160asi/a011270.rcw>
+#include <../lx2160asi/a050234.rcw>
 
 /* common PBI commands */
 #include <../lx2160asi/common.rcw>
diff --git a/lx2160aqds/FFFF_SSSS_PPPP_PPPP_PPPP_PPPP_RR_7_3_3/rcw_1900_600_2600_7_3_3_dspi.rcw b/lx2160aqds/FFFF_SSSS_PPPP_PPPP_PPPP_PPPP_RR_7_3_3/rcw_1900_600_2600_7_3_3_dspi.rcw
index 71d965a..eef347f 100644
--- a/lx2160aqds/FFFF_SSSS_PPPP_PPPP_PPPP_PPPP_RR_7_3_3/rcw_1900_600_2600_7_3_3_dspi.rcw
+++ b/lx2160aqds/FFFF_SSSS_PPPP_PPPP_PPPP_PPPP_RR_7_3_3/rcw_1900_600_2600_7_3_3_dspi.rcw
@@ -57,6 +57,7 @@ SRDS_DIV_PEX_S3=1
 
 /* Errata for PCIe controller */
 #include <../lx2160asi/a011270.rcw>
+#include <../lx2160asi/a050234.rcw>
 
 /* common PBI commands */
 #include <../lx2160asi/common.rcw>
diff --git a/lx2160aqds/FFGG_FFFF_PPPP_PPPP_PPPP_PPPP_RR_18_2_2/rcw_2000_700_2900_18_2_2.rcw b/lx2160aqds/FFGG_FFFF_PPPP_PPPP_PPPP_PPPP_RR_18_2_2/rcw_2000_700_2900_18_2_2.rcw
index 963f9dc..ce7c5e4 100644
--- a/lx2160aqds/FFGG_FFFF_PPPP_PPPP_PPPP_PPPP_RR_18_2_2/rcw_2000_700_2900_18_2_2.rcw
+++ b/lx2160aqds/FFGG_FFFF_PPPP_PPPP_PPPP_PPPP_RR_18_2_2/rcw_2000_700_2900_18_2_2.rcw
@@ -54,6 +54,7 @@ SRDS_DIV_PEX_S3=1
 
 /* Errata for PCIe controller */
 #include <../lx2160asi/a011270.rcw>
+#include <../lx2160asi/a050234.rcw>
 
 /* common PBI commands */
 #include <../lx2160asi/common.rcw>
diff --git a/lx2160aqds/FFGG_XXXX_PPPP_HHHHH_PPPP_PPPP_RR_19_5_2/rcw_2000_700_2600_19_5_2_sd.rcw b/lx2160aqds/FFGG_XXXX_PPPP_HHHHH_PPPP_PPPP_RR_19_5_2/rcw_2000_700_2600_19_5_2_sd.rcw
index 131dd28..5ad1190 100644
--- a/lx2160aqds/FFGG_XXXX_PPPP_HHHHH_PPPP_PPPP_RR_19_5_2/rcw_2000_700_2600_19_5_2_sd.rcw
+++ b/lx2160aqds/FFGG_XXXX_PPPP_HHHHH_PPPP_PPPP_RR_19_5_2/rcw_2000_700_2600_19_5_2_sd.rcw
@@ -60,6 +60,7 @@ blockcopy 0x08,0x00100000,0x1800a000,0x00020000
 
 /* Errata for PCIe controller */
 #include <../lx2160asi/a011270.rcw>
+#include <../lx2160asi/a050234.rcw>
 
 /* common PBI commands */
 #include <../lx2160asi/common.rcw>
diff --git a/lx2160aqds/LLLL_PPPP_PPPP_HHHH_PPPP_PPPP_RR_15_5_2/rcw_1900_600_2600_15_5_2.rcw b/lx2160aqds/LLLL_PPPP_PPPP_HHHH_PPPP_PPPP_RR_15_5_2/rcw_1900_600_2600_15_5_2.rcw
index 94a7883..3d9c43f 100644
--- a/lx2160aqds/LLLL_PPPP_PPPP_HHHH_PPPP_PPPP_RR_15_5_2/rcw_1900_600_2600_15_5_2.rcw
+++ b/lx2160aqds/LLLL_PPPP_PPPP_HHHH_PPPP_PPPP_RR_15_5_2/rcw_1900_600_2600_15_5_2.rcw
@@ -54,6 +54,7 @@ SRDS_DIV_PEX_S3=1
 
 /* Errata for PCIe controller */
 #include <../lx2160asi/a011270.rcw>
+#include <../lx2160asi/a050234.rcw>
 
 /* common PBI commands */
 #include <../lx2160asi/common.rcw>
diff --git a/lx2160aqds/LLLL_PPPP_PPPP_HHHH_PPPP_PPPP_RR_15_5_2/rcw_2000_700_2900_15_5_2.rcw b/lx2160aqds/LLLL_PPPP_PPPP_HHHH_PPPP_PPPP_RR_15_5_2/rcw_2000_700_2900_15_5_2.rcw
index 9121f1b..502e299 100644
--- a/lx2160aqds/LLLL_PPPP_PPPP_HHHH_PPPP_PPPP_RR_15_5_2/rcw_2000_700_2900_15_5_2.rcw
+++ b/lx2160aqds/LLLL_PPPP_PPPP_HHHH_PPPP_PPPP_RR_15_5_2/rcw_2000_700_2900_15_5_2.rcw
@@ -54,6 +54,7 @@ SRDS_DIV_PEX_S3=1
 
 /* Errata for PCIe controller */
 #include <../lx2160asi/a011270.rcw>
+#include <../lx2160asi/a050234.rcw>
 
 /* common PBI commands */
 #include <../lx2160asi/common.rcw>
diff --git a/lx2160aqds/PPPP_PPPP_PPPP_PPPP_PPPP_PPPP_RR_1_2_2/rcw_2000_700_2400_1_2_2.rcw b/lx2160aqds/PPPP_PPPP_PPPP_PPPP_PPPP_PPPP_RR_1_2_2/rcw_2000_700_2400_1_2_2.rcw
index 4cf0727..078704d 100644
--- a/lx2160aqds/PPPP_PPPP_PPPP_PPPP_PPPP_PPPP_RR_1_2_2/rcw_2000_700_2400_1_2_2.rcw
+++ b/lx2160aqds/PPPP_PPPP_PPPP_PPPP_PPPP_PPPP_RR_1_2_2/rcw_2000_700_2400_1_2_2.rcw
@@ -53,6 +53,7 @@ SRDS_DIV_PEX_S3=1
 
 /* Errata for PCIe controller */
 #include <../lx2160asi/a011270.rcw>
+#include <../lx2160asi/a050234.rcw>
 
 /* common PBI commands */
 #include <../lx2160asi/common.rcw>
diff --git a/lx2160aqds/PPPP_PPPP_PPPP_PPPP_PPPP_PPPP_RR_1_2_2/rcw_2000_700_2600_1_2_2.rcw b/lx2160aqds/PPPP_PPPP_PPPP_PPPP_PPPP_PPPP_RR_1_2_2/rcw_2000_700_2600_1_2_2.rcw
index 887141f..95041f0 100644
--- a/lx2160aqds/PPPP_PPPP_PPPP_PPPP_PPPP_PPPP_RR_1_2_2/rcw_2000_700_2600_1_2_2.rcw
+++ b/lx2160aqds/PPPP_PPPP_PPPP_PPPP_PPPP_PPPP_RR_1_2_2/rcw_2000_700_2600_1_2_2.rcw
@@ -53,6 +53,7 @@ SRDS_DIV_PEX_S3=1
 
 /* Errata for PCIe controller */
 #include <../lx2160asi/a011270.rcw>
+#include <../lx2160asi/a050234.rcw>
 
 /* common PBI commands */
 #include <../lx2160asi/common.rcw>
diff --git a/lx2160aqds/PPPP_PPPP_PPPP_PPPP_PPPP_PPPP_RR_1_2_2/rcw_2000_700_2900_1_2_2.rcw b/lx2160aqds/PPPP_PPPP_PPPP_PPPP_PPPP_PPPP_RR_1_2_2/rcw_2000_700_2900_1_2_2.rcw
index 77c6290..929db12 100644
--- a/lx2160aqds/PPPP_PPPP_PPPP_PPPP_PPPP_PPPP_RR_1_2_2/rcw_2000_700_2900_1_2_2.rcw
+++ b/lx2160aqds/PPPP_PPPP_PPPP_PPPP_PPPP_PPPP_RR_1_2_2/rcw_2000_700_2900_1_2_2.rcw
@@ -53,6 +53,7 @@ SRDS_DIV_PEX_S3=1
 
 /* Errata for PCIe controller */
 #include <../lx2160asi/a011270.rcw>
+#include <../lx2160asi/a050234.rcw>
 
 /* common PBI commands */
 #include <../lx2160asi/common.rcw>
diff --git a/lx2160aqds/README b/lx2160aqds/README
index 3e72208..87ef66a 100644
--- a/lx2160aqds/README
+++ b/lx2160aqds/README
@@ -212,3 +212,10 @@ Errata Workaround Implemented
 A-011270:
 	Software will read incorrect Function Dependency List for PF1
 	Workaround: Set FDL field in the SRIOV_FDL_DEVICE_ID1 to 0x1
+
+A-050234:
+	The default configuration of the elastic buffer thresholds in the PCI
+	Express controller is not optimal and can cause various link errors.
+	Workaround:
+	For Gen1 & Gen2, write 0x2008_1004 to PEXnCR0
+	For Gen3, write 0x0008_1004 to PEXnCR1
diff --git a/lx2160aqds/XXXX_XXXX_PSSS_PSSS_PPPP_PPPP_RR_20_11_3/rcw_2000_700_2400_20_11_3.rcw b/lx2160aqds/XXXX_XXXX_PSSS_PSSS_PPPP_PPPP_RR_20_11_3/rcw_2000_700_2400_20_11_3.rcw
index da10345..3516fe4 100644
--- a/lx2160aqds/XXXX_XXXX_PSSS_PSSS_PPPP_PPPP_RR_20_11_3/rcw_2000_700_2400_20_11_3.rcw
+++ b/lx2160aqds/XXXX_XXXX_PSSS_PSSS_PPPP_PPPP_RR_20_11_3/rcw_2000_700_2400_20_11_3.rcw
@@ -54,6 +54,7 @@ SRDS_DIV_PEX_S3=1
 
 /* Errata for PCIe controller */
 #include <../lx2160asi/a011270.rcw>
+#include <../lx2160asi/a050234.rcw>
 
 /* common PBI commands */
 #include <../lx2160asi/common.rcw>
diff --git a/lx2160ardb/IGGFF_PP_HHHH_RR_18_5_2/rcw_2000_700_2900_18_5_2.rcw b/lx2160ardb/IGGFF_PP_HHHH_RR_18_5_2/rcw_2000_700_2900_18_5_2.rcw
index 1107742..312a817 100644
--- a/lx2160ardb/IGGFF_PP_HHHH_RR_18_5_2/rcw_2000_700_2900_18_5_2.rcw
+++ b/lx2160ardb/IGGFF_PP_HHHH_RR_18_5_2/rcw_2000_700_2900_18_5_2.rcw
@@ -51,6 +51,7 @@ SRDS_DIV_PEX_S3=1
 
 /* Errata for PCIe controller */
 #include <../lx2160asi/a011270.rcw>
+#include <../lx2160asi/a050234.rcw>
 
 /* common PBI commands */
 #include <../lx2160asi/common.rcw>
diff --git a/lx2160ardb/README b/lx2160ardb/README
index 53aa015..95ca8be 100644
--- a/lx2160ardb/README
+++ b/lx2160ardb/README
@@ -70,3 +70,10 @@ Errata Workaround Implemented
 A-011270:
 	Software will read incorrect Function Dependency List for PF1
 	Workaround: Set FDL field in the SRIOV_FDL_DEVICE_ID1 to 0x1
+
+A-050234:
+	The default configuration of the elastic buffer thresholds in the PCI
+	Express controller is not optimal and can cause various link errors.
+	Workaround:
+	For Gen1 & Gen2, write 0x2008_1004 to PEXnCR0
+	For Gen3, write 0x0008_1004 to PEXnCR1
diff --git a/lx2160ardb/XGGFF_PP_HHHH_RR_19_5_2/rcw_1900_600_1600_19_5_2.rcw b/lx2160ardb/XGGFF_PP_HHHH_RR_19_5_2/rcw_1900_600_1600_19_5_2.rcw
index e30afb4..1623177 100644
--- a/lx2160ardb/XGGFF_PP_HHHH_RR_19_5_2/rcw_1900_600_1600_19_5_2.rcw
+++ b/lx2160ardb/XGGFF_PP_HHHH_RR_19_5_2/rcw_1900_600_1600_19_5_2.rcw
@@ -55,6 +55,7 @@ SRDS_DIV_PEX_S3=1
 
 /* Errata for PCIe controller */
 #include <../lx2160asi/a011270.rcw>
+#include <../lx2160asi/a050234.rcw>
 
 /* common PBI commands */
 #include <../lx2160asi/common.rcw>
diff --git a/lx2160ardb/XGGFF_PP_HHHH_RR_19_5_2/rcw_1900_600_2400_19_5_2.rcw b/lx2160ardb/XGGFF_PP_HHHH_RR_19_5_2/rcw_1900_600_2400_19_5_2.rcw
index 50b5179..8896ebb 100644
--- a/lx2160ardb/XGGFF_PP_HHHH_RR_19_5_2/rcw_1900_600_2400_19_5_2.rcw
+++ b/lx2160ardb/XGGFF_PP_HHHH_RR_19_5_2/rcw_1900_600_2400_19_5_2.rcw
@@ -55,6 +55,7 @@ SRDS_DIV_PEX_S3=3
 
 /* Errata for PCIe controller */
 #include <../lx2160asi/a011270.rcw>
+#include <../lx2160asi/a050234.rcw>
 
 /* common PBI commands */
 #include <../lx2160asi/common.rcw>
diff --git a/lx2160ardb/XGGFF_PP_HHHH_RR_19_5_2/rcw_1900_600_2600_19_5_2.rcw b/lx2160ardb/XGGFF_PP_HHHH_RR_19_5_2/rcw_1900_600_2600_19_5_2.rcw
index a30ea2c..97daea7 100644
--- a/lx2160ardb/XGGFF_PP_HHHH_RR_19_5_2/rcw_1900_600_2600_19_5_2.rcw
+++ b/lx2160ardb/XGGFF_PP_HHHH_RR_19_5_2/rcw_1900_600_2600_19_5_2.rcw
@@ -52,6 +52,7 @@ SRDS_DIV_PEX_S3=1
 
 /* Errata for PCIe controller */
 #include <../lx2160asi/a011270.rcw>
+#include <../lx2160asi/a050234.rcw>
 
 /* common PBI commands */
 #include <../lx2160asi/common.rcw>
diff --git a/lx2160ardb/XGGFF_PP_HHHH_RR_19_5_2/rcw_1900_600_2600_19_5_2_sd.rcw b/lx2160ardb/XGGFF_PP_HHHH_RR_19_5_2/rcw_1900_600_2600_19_5_2_sd.rcw
index 273279f..29ce8c2 100644
--- a/lx2160ardb/XGGFF_PP_HHHH_RR_19_5_2/rcw_1900_600_2600_19_5_2_sd.rcw
+++ b/lx2160ardb/XGGFF_PP_HHHH_RR_19_5_2/rcw_1900_600_2600_19_5_2_sd.rcw
@@ -59,6 +59,7 @@ blockcopy 0x08,0x00100000,0x1800a000,0x00020000
 
 /* Errata for PCIe controller */
 #include <../lx2160asi/a011270.rcw>
+#include <../lx2160asi/a050234.rcw>
 
 /* common PBI commands */
 #include <../lx2160asi/common.rcw>
diff --git a/lx2160ardb/XGGFF_PP_HHHH_RR_19_5_2/rcw_2000_700_2400_19_5_2.rcw b/lx2160ardb/XGGFF_PP_HHHH_RR_19_5_2/rcw_2000_700_2400_19_5_2.rcw
index a02c92c..bdbf1fd 100644
--- a/lx2160ardb/XGGFF_PP_HHHH_RR_19_5_2/rcw_2000_700_2400_19_5_2.rcw
+++ b/lx2160ardb/XGGFF_PP_HHHH_RR_19_5_2/rcw_2000_700_2400_19_5_2.rcw
@@ -55,6 +55,7 @@ SRDS_DIV_PEX_S3=3
 
 /* Errata for PCIe controller */
 #include <../lx2160asi/a011270.rcw>
+#include <../lx2160asi/a050234.rcw>
 
 /* common PBI commands */
 #include <../lx2160asi/common.rcw>
diff --git a/lx2160ardb/XGGFF_PP_HHHH_RR_19_5_2/rcw_2000_700_2900_19_5_2.rcw b/lx2160ardb/XGGFF_PP_HHHH_RR_19_5_2/rcw_2000_700_2900_19_5_2.rcw
index 984a6d9..fa4570d 100644
--- a/lx2160ardb/XGGFF_PP_HHHH_RR_19_5_2/rcw_2000_700_2900_19_5_2.rcw
+++ b/lx2160ardb/XGGFF_PP_HHHH_RR_19_5_2/rcw_2000_700_2900_19_5_2.rcw
@@ -55,6 +55,7 @@ SRDS_DIV_PEX_S3=1
 
 /* Errata for PCIe controller */
 #include <../lx2160asi/a011270.rcw>
+#include <../lx2160asi/a050234.rcw>
 
 /* Modify FlexSPI Clock Divisor value */
 #include <../lx2160asi/flexspi_divisor_28.rcw>
diff --git a/lx2160ardb/XGGFF_PP_HHHH_RR_19_5_2/rcw_2000_700_2900_19_5_2_sd.rcw b/lx2160ardb/XGGFF_PP_HHHH_RR_19_5_2/rcw_2000_700_2900_19_5_2_sd.rcw
index 17abb39..b51072c 100644
--- a/lx2160ardb/XGGFF_PP_HHHH_RR_19_5_2/rcw_2000_700_2900_19_5_2_sd.rcw
+++ b/lx2160ardb/XGGFF_PP_HHHH_RR_19_5_2/rcw_2000_700_2900_19_5_2_sd.rcw
@@ -59,6 +59,7 @@ blockcopy 0x08,0x00100000,0x1800a000,0x00015000
 
 /* Errata for PCIe controller */
 #include <../lx2160asi/a011270.rcw>
+#include <../lx2160asi/a050234.rcw>
 
 /* common PBI commands */
 #include <../lx2160asi/common.rcw>
diff --git a/lx2160asi/a050234.rcw b/lx2160asi/a050234.rcw
new file mode 100644
index 0000000..72a40e4
--- /dev/null
+++ b/lx2160asi/a050234.rcw
@@ -0,0 +1,33 @@
+/*
+ * The default configuration of the elastic buffer thresholds in the PCI
+ * Express controller is not optimal and can cause various link errors.
+ */
+
+.pbi
+write 0x1ea1200,0x20081004
+write 0x1ea1240,0x20081004
+write 0x1ea1280,0x20081004
+write 0x1ea12C0,0x20081004
+write 0x1ea1204,0x00081004
+write 0x1ea1244,0x00081004
+write 0x1ea1284,0x00081004
+write 0x1ea12C4,0x00081004
+
+write 0x1eb1200,0x20081004
+write 0x1eb1240,0x20081004
+write 0x1eb1280,0x20081004
+write 0x1eb12C0,0x20081004
+write 0x1eb1204,0x00081004
+write 0x1eb1244,0x00081004
+write 0x1eb1284,0x00081004
+write 0x1eb12C4,0x00081004
+
+write 0x1ec1200,0x20081004
+write 0x1ec1240,0x20081004
+write 0x1ec1280,0x20081004
+write 0x1ec12C0,0x20081004
+write 0x1ec1204,0x00081004
+write 0x1ec1244,0x00081004
+write 0x1ec1284,0x00081004
+write 0x1ec12C4,0x00081004
+.end
-- 
2.17.1

